File: ios/Sendro/Views/Theme.swift
//
//  Theme.swift
//  Sendro
//
//  The new Sendro design language: near-black canvas with a top iris glow,
//  glass cards, iris (periwinkle) energy accent + brand teal for "verified",
//  SF Pro for prose and SF Mono for data labels.
//
//  Iris values are oklch() from the design prototype, gamut-mapped to sRGB:
//    oklch(0.72 0.18 265) -> #78A1FF   (primary iris)
//    oklch(0.78 0.15 265) -> #95B6FF   (soft iris — labels, links)
//    oklch(0.82 0.15 265) -> #A8C4FF   (bright iris — glows)
//

import SwiftUI

// MARK: - Palette

enum Theme {

    static let bg          = Color(red: 0x07 / 255, green: 0x08 / 255, blue: 0x0B / 255)
    static let bgGlow      = Color(red: 0x16 / 255, green: 0x1A / 255, blue: 0x2E / 255)

    static let iris        = Color(red: 0x78 / 255, green: 0xA1 / 255, blue: 0xFF / 255)
    static let irisSoft    = Color(red: 0x95 / 255, green: 0xB6 / 255, blue: 0xFF / 255)
    static let irisBright  = Color(red: 0xA8 / 255, green: 0xC4 / 255, blue: 0xFF / 255)
    static let teal        = Color(red: 0x37 / 255, green: 0xE6 / 255, blue: 0xC4 / 255)
    /// Beam gradient ends — these three are the app icon's stroke gradient
    /// (see scripts/generate_icons.py): #1FB78F → #37E6C4 → #6BF2D6.
    static let tealDeep    = Color(red: 0x1F / 255, green: 0xB7 / 255, blue: 0x8F / 255)
    static let tealBright  = Color(red: 0x6B / 255, green: 0xF2 / 255, blue: 0xD6 / 255)
    /// The icon tile's background gradient: #151A21 → #0A0C10.
    static let markTileTop    = Color(red: 0x15 / 255, green: 0x1A / 255, blue: 0x21 / 255)
    static let markTileBottom = Color(red: 0x0A / 255, green: 0x0C / 255, blue: 0x10 / 255)
    static let danger      = Color(red: 1.0,        green: 0x78 / 255, blue: 0x78 / 255)
    static let warn        = Color(red: 1.0,        green: 0xB8 / 255, blue: 0x6B / 255)

    /// Ink used on top of iris / teal filled buttons.
    static let onAccent    = Color(red: 0x0A / 255, green: 0x0B / 255, blue: 0x14 / 255)

    static let textPrimary = Color(red: 0xF5 / 255, green: 0xF6 / 255, blue: 0xFA / 255)
    /// Base for secondary text — use with .opacity().
    static let textBase    = Color(red: 0xF2 / 255, green: 0xF3 / 255, blue: 0xF7 / 255)

    static var textSecondary: Color { textBase.opacity(0.55) }
    static var textTertiary: Color  { textBase.opacity(0.42) }
    static var textFaint: Color     { textBase.opacity(0.35) }

    // MARK: Type

    static func sans(_ size: CGFloat, _ weight: Font.Weight = .regular) -> Font {
        .system(size: size, weight: weight, design: .default)
    }

    static func mono(_ size: CGFloat, _ weight: Font.Weight = .regular) -> Font {
        .system(size: size, weight: weight, design: .monospaced)
    }
}

// MARK: - Canvas background

/// Near-black canvas with the radial iris glow bleeding in from above.
///
/// LAYOUT NOTE: the 520pt glow disc must NOT be a layout child — a fixed
/// 520pt-wide sibling inflates the parent ZStack's union size past any
/// iPhone width (393pt) and the whole UI gets centered + clipped on both
/// edges. It lives in an .overlay (which never affects layout) and is
/// clipped, so this view always reports exactly the proposed size.
struct SendroBackground: View {
    var body: some View {
        Theme.bg
            .ignoresSafeArea()
            .overlay(alignment: .top) {
                RadialGradient(colors: [Theme.iris.opacity(0.22),
                                        Theme.iris.opacity(0.06),
                                        .clear],
                               center: .center,
                               startRadius: 0,
                               endRadius: 260)
                    .frame(width: 520, height: 520)
                    .offset(y: -220)
                    .allowsHitTesting(false)
            }
    }
}

// MARK: - Glass surfaces

/// Big hero glass card (incoming offer, sheet panels).
struct GlassCard: ViewModifier {
    var cornerRadius: CGFloat = 26

    func body(content: Content) -> some View {
        content
            .background(
                RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
                    .fill(LinearGradient(colors: [Color.white.opacity(0.09),
                                                  Color.white.opacity(0.035)],
                                         startPoint: .topLeading,
                                         endPoint: .bottomTrailing))
                    .background(.ultraThinMaterial,
                                in: RoundedRectangle(cornerRadius: cornerRadius, style: .continuous))
            )
            .overlay(
                RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
                    .strokeBorder(Color.white.opacity(0.13), lineWidth: 0.5)
            )
            .shadow(color: .black.opacity(0.5), radius: 30, x: 0, y: 20)
    }
}

/// Quiet list-row glass.
struct GlassRow: ViewModifier {
    var cornerRadius: CGFloat = 18
    var fillOpacity: Double = 0.045
    var borderOpacity: Double = 0.07

    func body(content: Content) -> some View {
        content
            .background(
                RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
                    .fill(Color.white.opacity(fillOpacity))
            )
            .overlay(
                RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
                    .strokeBorder(Color.white.opacity(borderOpacity), lineWidth: 0.5)
            )
    }
}

extension View {
    func glassCard(cornerRadius: CGFloat = 26) -> some View {
        modifier(GlassCard(cornerRadius: cornerRadius))
    }

    func glassRow(cornerRadius: CGFloat = 18,
                  fillOpacity: Double = 0.045,
                  borderOpacity: Double = 0.07) -> some View {
        modifier(GlassRow(cornerRadius: cornerRadius,
                          fillOpacity: fillOpacity,
                          borderOpacity: borderOpacity))
    }
}

// MARK: - Buttons

/// Scale-on-press, like the prototype's style-active transforms.
struct PressableButtonStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed ? 0.97 : 1.0)
            .animation(.spring(response: 0.25, dampingFraction: 0.7),
                       value: configuration.isPressed)
    }
}

/// Filled primary action (iris by default).
struct AccentPillLabel: View {
    let title: String
    var color: Color = Theme.iris
    var height: CGFloat = 48

    var body: some View {
        Text(title)
            .font(Theme.sans(15.5, .semibold))
            .foregroundColor(Theme.onAccent)
            .lineLimit(1)
            .minimumScaleFactor(0.75)
            .padding(.horizontal, 8)
            .frame(maxWidth: .infinity)
            .frame(height: height)
            .background(RoundedRectangle(cornerRadius: 16, style: .continuous).fill(color))
            .shadow(color: color.opacity(0.4), radius: 14, x: 0, y: 8)
    }
}

/// Quiet glass action.
struct GhostPillLabel: View {
    let title: String
    var textColor: Color = Theme.textBase.opacity(0.7)
    var height: CGFloat = 48

    var body: some View {
        Text(title)
            .font(Theme.sans(15.5, .medium))
            .foregroundColor(textColor)
            .lineLimit(1)
            .minimumScaleFactor(0.75)
            .padding(.horizontal, 8)
            .frame(maxWidth: .infinity)
            .frame(height: height)
            .glassRow(cornerRadius: 16, fillOpacity: 0.07, borderOpacity: 0.1)
    }
}

// MARK: - Labels

/// Uppercase tracked mono section label ("INCOMING", "RECENT", …).
struct SectionTag: View {
    let text: String
    var color: Color = Theme.textFaint

    var body: some View {
        Text(text.uppercased())
            .font(Theme.mono(10.5, .medium))
            .tracking(1.8)
            .foregroundColor(color)
    }
}

// MARK: - File type badge

/// Hatched square with the uppercase extension, standing in for previews.
struct FileBadge: View {
    let fileName: String
    var side: CGFloat = 44
    var cornerRadius: CGFloat = 13

    private var ext: String {
        let e = (fileName as NSString).pathExtension.uppercased()
        return e.isEmpty ? "FILE" : String(e.prefix(4))
    }

    var body: some View {
        ZStack {
            HatchPattern()
                .clipShape(RoundedRectangle(cornerRadius: cornerRadius, style: .continuous))
            Text(ext)
                .font(Theme.mono(side < 40 ? 8.5 : 9.5, .semibold))
                .foregroundColor(Theme.textBase.opacity(0.72))
                .lineLimit(1)
                .minimumScaleFactor(0.6)
                .padding(.horizontal, 3)
        }
        .frame(width: side, height: side)
        .overlay(
            RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
                .strokeBorder(Color.white.opacity(0.1), lineWidth: 0.5)
        )
    }
}

/// Cheap 135° repeating hatch, like the prototype's repeating-linear-gradient.
struct HatchPattern: View {
    var body: some View {
        GeometryReader { geo in
            let size = max(geo.size.width, geo.size.height)
            Canvas { context, _ in
                let stripe: CGFloat = 5
                context.fill(Path(CGRect(origin: .zero,
                                         size: CGSize(width: size, height: size))),
                             with: .color(Color.white.opacity(0.04)))
                var x: CGFloat = -size
                while x < size * 2 {
                    var path = Path()
                    path.move(to: CGPoint(x: x, y: 0))
                    path.addLine(to: CGPoint(x: x + size, y: size))
                    context.stroke(path,
                                   with: .color(Color.white.opacity(0.07)),
                                   lineWidth: stripe)
                    x += stripe * 2
                }
            }
        }
    }
}

// MARK: - Beam mark (must match the shipped app icon)

/// The Sendro beam: the exact path the icon pipeline draws
/// (scripts/generate_icons.py), `M 252 708 C 560 708, 452 316, 700 316`,
/// expressed in the icon's 1024 × 1024 design space and scaled into the
/// view's rect. Stroke it 88/1024 wide with round caps.
struct BeamMarkShape: Shape {

    /// The icon's design canvas — every constant below is in this space.
    static let canvas: CGFloat = 1024

    func path(in rect: CGRect) -> Path {
        let unit = min(rect.width, rect.height) / Self.canvas
        // Centre the 1024-square inside a non-square rect so the geometry
        // never skews.
        let originX = rect.minX + (rect.width - Self.canvas * unit) / 2
        let originY = rect.minY + (rect.height - Self.canvas * unit) / 2
        func point(_ x: CGFloat, _ y: CGFloat) -> CGPoint {
            CGPoint(x: originX + x * unit, y: originY + y * unit)
        }
        var path = Path()
        path.move(to: point(252, 708))
        path.addCurve(to: point(700, 316),
                      control1: point(560, 708),
                      control2: point(452, 316))
        return path
    }
}

/// The app mark, pixel-faithful to the home-screen icon: the #151A21→#0A0C10
/// tile with its teal ambient glow, the gradient beam (#1FB78F → #37E6C4 →
/// #6BF2D6), the bright destination dot at (806, 316) r 42, and the 6/1024
/// hairline edge — all in the icon's 1024 space, scaled to `side`.
struct BeamMark: View {

    var side: CGFloat = 28

    /// One unit of the icon's 1024-space, in points.
    private var unit: CGFloat { side / BeamMarkShape.canvas }
    private var corner: CGFloat { 228 * unit }

    /// Offset of the destination dot's centre from the tile's centre.
    private var dotOffset: CGSize {
        CGSize(width: (806 - 512) * unit, height: (316 - 512) * unit)
    }

    private var beamGradient: LinearGradient {
        LinearGradient(gradient: Gradient(stops: [
            Gradient.Stop(color: Theme.tealDeep, location: 0),
            Gradient.Stop(color: Theme.teal, location: 0.62),
            Gradient.Stop(color: Theme.tealBright, location: 1)
        ]),
        startPoint: UnitPoint(x: 300 / BeamMarkShape.canvas, y: 712 / BeamMarkShape.canvas),
        endPoint: UnitPoint(x: 756 / BeamMarkShape.canvas, y: 312 / BeamMarkShape.canvas))
    }

    var body: some View {
        ZStack {
            // Tile: vertical gradient + the faint teal ambient wash.
            LinearGradient(colors: [Theme.markTileTop, Theme.markTileBottom],
                           startPoint: .top, endPoint: .bottom)
                .overlay(
                    RadialGradient(gradient: Gradient(stops: [
                        Gradient.Stop(color: Theme.teal.opacity(0.10), location: 0),
                        Gradient.Stop(color: Theme.teal.opacity(0.03), location: 0.55),
                        Gradient.Stop(color: Theme.teal.opacity(0), location: 1)
                    ]),
                    center: UnitPoint(x: 0.5, y: 330 / BeamMarkShape.canvas),
                    startRadius: 0,
                    endRadius: 620 * unit)
                )

            // Glow pass (stroke-width 92 @ 0.32 + dot r 46 @ 0.35, blurred).
            BeamMarkShape()
                .stroke(Theme.teal.opacity(0.32),
                        style: StrokeStyle(lineWidth: 92 * unit, lineCap: .round))
                .blur(radius: 22 * unit)
            Circle()
                .fill(Theme.teal.opacity(0.35))
                .frame(width: 92 * unit, height: 92 * unit)
                .offset(x: dotOffset.width, y: dotOffset.height)
                .blur(radius: 22 * unit)

            // The mark itself.
            BeamMarkShape()
                .stroke(beamGradient,
                        style: StrokeStyle(lineWidth: 88 * unit, lineCap: .round))
            Circle()
                .fill(Theme.tealBright)
                .frame(width: 84 * unit, height: 84 * unit)
                .offset(x: dotOffset.width, y: dotOffset.height)
        }
        .frame(width: side, height: side)
        .clipShape(RoundedRectangle(cornerRadius: corner, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: corner, style: .continuous)
                .strokeBorder(Color.white.opacity(0.055), lineWidth: 6 * unit)
        )
    }
}

// MARK: - Pulse dot

/// Small status dot with an expanding pulse ring when active.
struct PulseDot: View {
    var color: Color = Theme.teal
    var active: Bool = true
    var side: CGFloat = 7

    var body: some View {
        ZStack {
            if active {
                PulseRing(color: color, side: side)
            }
            Circle()
                .fill(active ? color : Theme.textBase.opacity(0.3))
                .frame(width: side, height: side)
                .shadow(color: active ? color.opacity(0.8) : .clear, radius: 6)
        }
    }
}

/// The expanding ring; owns its animation so it restarts on insertion.
private struct PulseRing: View {
    let color: Color
    let side: CGFloat

    @State private var pulsing = false

    var body: some View {
        Circle()
            .fill(color)
            .frame(width: side, height: side)
            .scaleEffect(pulsing ? 2.4 : 1.0)
            .opacity(pulsing ? 0 : 0.9)
            .onAppear {
                withAnimation(.easeOut(duration: 2.2).repeatForever(autoreverses: false)) {
                    pulsing = true
                }
            }
    }
}

File: android/app/src/main/java/com/sendro/android/ui/theme/Theme.kt
package com.sendro.android.ui.theme

import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Typography
import androidx.compose.material3.darkColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.clipPath
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * The Sendro design language, ported token for token from
 * ios/Sendro/Views/Theme.swift.
 *
 * Near-black canvas with a top iris glow, glass cards, iris (periwinkle) as
 * the energy accent and the brand teal for "verified / online". System sans
 * for prose, monospace for the technical micro-labels.
 *
 * Iris values are oklch() from the design prototype, gamut-mapped to sRGB:
 *   oklch(0.72 0.18 265) -> #78A1FF   (primary iris)
 *   oklch(0.78 0.15 265) -> #95B6FF   (soft iris — labels, links)
 *   oklch(0.82 0.15 265) -> #A8C4FF   (bright iris — glows)
 */
object Sendro {

    // -- Palette ------------------------------------------------------------

    val bg = Color(0xFF07080B)
    val bgGlow = Color(0xFF161A2E)

    val iris = Color(0xFF78A1FF)
    val irisSoft = Color(0xFF95B6FF)
    val irisBright = Color(0xFFA8C4FF)

    val teal = Color(0xFF37E6C4)

    /** Beam gradient ends — the app icon's stroke gradient. */
    val tealDeep = Color(0xFF1FB78F)
    val tealBright = Color(0xFF6BF2D6)

    /** The icon tile's background gradient. */
    val markTileTop = Color(0xFF151A21)
    val markTileBottom = Color(0xFF0A0C10)

    val danger = Color(0xFFFF7878)
    val warn = Color(0xFFFFB86B)

    /** Ink used on top of iris / teal filled buttons. */
    val onAccent = Color(0xFF0A0B14)

    val textPrimary = Color(0xFFF5F6FA)

    /** Base for secondary text — always used with an alpha from the ramp. */
    val textBase = Color(0xFFF2F3F7)

    /**
     * The opacity ramp. On a TV the low end of it is simply invisible from a
     * sofa, so the alpha is lifted towards a floor — the *relationship*
     * between the three steps survives, the illegibility does not.
     */
    val textSecondary: Color
        @Composable get() = textBase.copy(alpha = rampAlpha(0.55f))
    val textTertiary: Color
        @Composable get() = textBase.copy(alpha = rampAlpha(0.42f))
    val textFaint: Color
        @Composable get() = textBase.copy(alpha = rampAlpha(0.35f))

    /** Static variants for the rare non-composable use (theme construction). */
    val textSecondaryStatic = textBase.copy(alpha = 0.55f)

    @Composable
    private fun rampAlpha(base: Float): Float {
        val floor = LocalDeviceProfile.current.minTextAlpha
        return if (floor <= 0f) base else maxOf(base, floor - (0.55f - base) * 0.35f)
    }

    // -- Type ---------------------------------------------------------------

    /**
     * Body/prose type.
     *
     * These are `@Composable` so they can read [LocalDeviceProfile]: on a TV
     * every size is scaled and body text gets a hard floor, without a single
     * call site changing. Every use is already inside a composable.
     */
    @Composable
    fun sans(size: Float, weight: FontWeight = FontWeight.Normal): TextStyle {
        val scaled = scaleType(size, LocalDeviceProfile.current)
        return TextStyle(
            fontFamily = FontFamily.Default,
            fontSize = scaled.sp,
            fontWeight = weight,
            lineHeight = (scaled * 1.32f).sp,
        )
    }

    /**
     * The technical voice: sizes, hashes, IPs, section tags. Monospace is the
     * one typographic signal that separates "data" from "prose" in Sendro,
     * and it is the same on both platforms.
     */
    @Composable
    fun mono(
        size: Float,
        weight: FontWeight = FontWeight.Normal,
        tracking: Float = 0f,
    ): TextStyle {
        val scaled = scaleType(size, LocalDeviceProfile.current)
        return TextStyle(
            fontFamily = FontFamily.Monospace,
            fontSize = scaled.sp,
            fontWeight = weight,
            letterSpacing = tracking.sp,
            lineHeight = (scaled * 1.3f).sp,
        )
    }

    /**
     * Ten-foot type: 1.3x everywhere, and a hard 18sp floor for anything that
     * was already body-sized. Micro-labels (the 9–11sp mono tags) are scaled
     * but not floored — flooring them too would turn a tracked caption into a
     * second headline.
     */
    private fun scaleType(size: Float, profile: DeviceProfile): Float {
        if (!profile.isTv) return size
        val scaled = size * profile.typeScale
        return if (size >= 13f) maxOf(scaled, 18f) else scaled
    }

    // -- Shapes -------------------------------------------------------------

    val cardShape: Shape = RoundedCornerShape(26.dp)
    val rowShape: Shape = RoundedCornerShape(18.dp)
    val pillShape: Shape = RoundedCornerShape(16.dp)
}

/**
 * Material3 is present as a base (Switch, TextField, ripples) but every colour
 * is Sendro's. Nothing in the app reads a Material token expecting Material
 * defaults.
 */
@Composable
fun SendroTheme(
    profile: DeviceProfile = rememberDeviceProfile(),
    content: @Composable () -> Unit,
) {
    val colors = darkColorScheme(
        primary = Sendro.iris,
        onPrimary = Sendro.onAccent,
        secondary = Sendro.teal,
        onSecondary = Sendro.onAccent,
        background = Sendro.bg,
        onBackground = Sendro.textPrimary,
        surface = Sendro.bg,
        onSurface = Sendro.textPrimary,
        surfaceVariant = Color.White.copy(alpha = 0.06f),
        onSurfaceVariant = Sendro.textSecondaryStatic,
        error = Sendro.danger,
        onError = Sendro.onAccent,
        outline = Color.White.copy(alpha = 0.12f),
    )
    // The profile has to be in scope BEFORE the typography is built, because
    // Sendro.sans/mono read it to decide the TV type scale.
    CompositionLocalProvider(LocalDeviceProfile provides profile) {
        MaterialTheme(
            colorScheme = colors,
            typography = Typography(
                bodyLarge = Sendro.sans(15f),
                bodyMedium = Sendro.sans(13.5f),
                labelLarge = Sendro.sans(14f, FontWeight.Medium),
            ),
            content = content,
        )
    }
}

// ---------------------------------------------------------------------------
// Canvas background
// ---------------------------------------------------------------------------

/**
 * Near-black canvas with the radial iris glow bleeding in from above.
 *
 * The glow is drawn, not laid out: a 520dp child would inflate the parent's
 * size past any phone width. `drawBehind` never affects layout.
 */
@Composable
fun SendroBackground(modifier: Modifier = Modifier, content: @Composable BoxScope.() -> Unit) {
    Box(
        modifier = modifier
            .fillMaxSize()
            .background(Sendro.bg)
            .drawBehind {
                val radius = 260.dp.toPx()
                drawCircle(
                    brush = Brush.radialGradient(
                        colorStops = arrayOf(
                            0f to Sendro.iris.copy(alpha = 0.22f),
                            0.5f to Sendro.iris.copy(alpha = 0.06f),
                            1f to Color.Transparent,
                        ),
                        center = Offset(size.width / 2f, -40.dp.toPx()),
                        radius = radius,
                    ),
                    radius = radius,
                    center = Offset(size.width / 2f, -40.dp.toPx()),
                )
            },
        content = content,
    )
}

// ---------------------------------------------------------------------------
// Glass surfaces
// ---------------------------------------------------------------------------

/**
 * Big hero glass card (incoming offer, sheet panels).
 *
 * NOTE: there is no backdrop blur. Compose's `Modifier.blur` is a RenderEffect
 * that only exists from API 31 and blurs the *content*, not what is behind it
 * — there is no `.ultraThinMaterial` equivalent below 31. The translucent
 * white gradient carries the same read on the dark canvas, and looks identical
 * on every supported device instead of good on new ones and flat on old ones.
 */
@Composable
fun Modifier.glassCard(
    cornerRadius: Dp = 26.dp,
    shadow: Dp = 20.dp,
): Modifier {
    val profile = LocalDeviceProfile.current
    val shape = RoundedCornerShape(cornerRadius)
    return this
        .shadow(
            elevation = shadow,
            shape = shape,
            clip = false,
            ambientColor = Color.Black,
            spotColor = Color.Black,
        )
        .clip(shape)
        .background(
            brush = Brush.linearGradient(
                colors = listOf(
                    Color.White.copy(alpha = 0.09f + profile.glassFillBoost),
                    Color.White.copy(alpha = 0.035f + profile.glassFillBoost * 0.8f),
                ),
            ),
        )
        .border(
            width = if (profile.isTv) 1.dp else 0.5.dp,
            color = Color.White.copy(alpha = 0.13f + profile.glassBorderBoost),
            shape = shape,
        )
}

/** Quiet list-row glass. */
@Composable
fun Modifier.glassRow(
    cornerRadius: Dp = 18.dp,
    fillAlpha: Float = 0.045f,
    borderAlpha: Float = 0.07f,
): Modifier {
    val profile = LocalDeviceProfile.current
    val shape = RoundedCornerShape(cornerRadius)
    return this
        .clip(shape)
        .background(Color.White.copy(alpha = fillAlpha + profile.glassFillBoost))
        .border(
            width = if (profile.isTv) 1.dp else 0.5.dp,
            color = Color.White.copy(alpha = borderAlpha + profile.glassBorderBoost),
            shape = shape,
        )
}

/**
 * The D-pad focus indicator.
 *
 * Drawn, not laid out: a border/padding change on focus would reflow the whole
 * row every time the remote moves. `drawWithContent` paints an iris ring plus a
 * soft outer glow just outside the element's bounds, which costs nothing and
 * never shifts anything.
 *
 * It is deliberately NOT colour-only — the ring is a shape change and it is
 * paired with a scale in [com.sendro.android.ui.components.Pressable], so it
 * survives a colour-blind viewer and a badly calibrated TV panel alike.
 */
fun Modifier.sendroFocusRing(
    focused: Boolean,
    cornerRadius: Dp,
    ringWidth: Dp,
    color: Color = Sendro.irisBright,
): Modifier = this.drawWithContent {
    drawContent()
    if (!focused) return@drawWithContent
    val stroke = ringWidth.toPx()
    val inset = stroke * 1.5f
    val radius = cornerRadius.toPx() + inset
    // Soft outer glow first, then the crisp ring on top.
    drawRoundRect(
        color = color.copy(alpha = 0.22f),
        topLeft = Offset(-inset * 2f, -inset * 2f),
        size = Size(size.width + inset * 4f, size.height + inset * 4f),
        cornerRadius = CornerRadius(radius + inset, radius + inset),
        style = Stroke(width = stroke * 2.4f),
    )
    drawRoundRect(
        color = color,
        topLeft = Offset(-inset, -inset),
        size = Size(size.width + inset * 2f, size.height + inset * 2f),
        cornerRadius = CornerRadius(radius, radius),
        style = Stroke(width = stroke),
    )
}

// ---------------------------------------------------------------------------
// The beam mark (must match the shipped app icon)
// ---------------------------------------------------------------------------

/**
 * The Sendro beam, pixel-faithful to the home-screen icon.
 *
 * Every constant is in the icon's 1024x1024 design space
 * (scripts/generate_icons.py) and scaled into the view:
 *
 *   beam   M 252 708 C 560 708, 452 316, 700 316   stroke 88, round caps
 *          gradient #1FB78F -> #37E6C4 -> #6BF2D6 along (300,712)->(756,312)
 *   dot    r 42 at (806, 316), #6BF2D6
 *   tile   #151A21 -> #0A0C10 vertical, corner radius 228/1024
 *   edge   6/1024 hairline at 5.5% white
 */
private const val ICON_CANVAS = 1024f

/** The bare beam path in the icon's 1024-space, scaled by [unit]. */
private fun beamPath(unit: Float, originX: Float, originY: Float): Path {
    fun x(v: Float) = originX + v * unit
    fun y(v: Float) = originY + v * unit
    return Path().apply {
        moveTo(x(252f), y(708f))
        cubicTo(x(560f), y(708f), x(452f), y(316f), x(700f), y(316f))
    }
}

@Composable
fun BeamMark(side: Dp = 28.dp, modifier: Modifier = Modifier) {
    val corner = side * (228f / ICON_CANVAS)
    Canvas(
        modifier = modifier
            .size(side)
            .clip(RoundedCornerShape(corner))
            .border(
                width = side * (6f / ICON_CANVAS),
                color = Color.White.copy(alpha = 0.055f),
                shape = RoundedCornerShape(corner),
            ),
    ) {
        drawBeamMark(this)
    }
}

/** Split out so both [BeamMark] and any decorative use share one drawing. */
private fun drawBeamMark(scope: DrawScope) = with(scope) {
    val unit = minOf(size.width, size.height) / ICON_CANVAS
    val originX = (size.width - ICON_CANVAS * unit) / 2f
    val originY = (size.height - ICON_CANVAS * unit) / 2f

    fun px(v: Float) = originX + v * unit
    fun py(v: Float) = originY + v * unit

    // Tile: vertical gradient + the faint teal ambient wash.
    drawRect(
        brush = Brush.verticalGradient(
            listOf(Sendro.markTileTop, Sendro.markTileBottom),
            startY = originY,
            endY = originY + ICON_CANVAS * unit,
        ),
    )
    drawCircle(
        brush = Brush.radialGradient(
            colorStops = arrayOf(
                0f to Sendro.teal.copy(alpha = 0.10f),
                0.55f to Sendro.teal.copy(alpha = 0.03f),
                1f to Color.Transparent,
            ),
            center = Offset(px(512f), py(330f)),
            radius = 620f * unit,
        ),
        radius = 620f * unit,
        center = Offset(px(512f), py(330f)),
    )

    val path = beamPath(unit, originX, originY)
    val gradient = Brush.linearGradient(
        colorStops = arrayOf(
            0f to Sendro.tealDeep,
            0.62f to Sendro.teal,
            1f to Sendro.tealBright,
        ),
        start = Offset(px(300f), py(712f)),
        end = Offset(px(756f), py(312f)),
    )

    // Glow pass (the icon draws stroke-width 92 at 0.32 alpha plus a soft dot;
    // a real blur is not available in a DrawScope, so the glow is approximated
    // by a wider, softer stroke underneath — the same read at icon sizes).
    drawPath(
        path = path,
        color = Sendro.teal.copy(alpha = 0.20f),
        style = Stroke(width = 116f * unit, cap = StrokeCap.Round),
    )
    drawCircle(
        color = Sendro.teal.copy(alpha = 0.22f),
        radius = 62f * unit,
        center = Offset(px(806f), py(316f)),
    )

    // The mark itself.
    drawPath(
        path = path,
        brush = gradient,
        style = Stroke(width = 88f * unit, cap = StrokeCap.Round),
    )
    drawCircle(
        color = Sendro.tealBright,
        radius = 42f * unit,
        center = Offset(px(806f), py(316f)),
    )
}

// ---------------------------------------------------------------------------
// Small primitives
// ---------------------------------------------------------------------------

/** Status dot with an expanding pulse ring when active. */
@Composable
fun PulseDot(
    color: Color = Sendro.teal,
    active: Boolean = true,
    side: Dp = 7.dp,
    modifier: Modifier = Modifier,
) {
    val transition = rememberInfiniteTransition(label = "pulse")
    val scale by transition.animateFloat(
        initialValue = 1f,
        targetValue = 2.4f,
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = 2200),
            repeatMode = RepeatMode.Restart,
        ),
        label = "pulseScale",
    )
    val alpha by transition.animateFloat(
        initialValue = 0.9f,
        targetValue = 0f,
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = 2200),
            repeatMode = RepeatMode.Restart,
        ),
        label = "pulseAlpha",
    )
    Canvas(modifier = modifier.size(side * 2.6f)) {
        val centre = Offset(size.width / 2f, size.height / 2f)
        val radius = side.toPx() / 2f
        if (active) {
            drawCircle(color = color.copy(alpha = alpha), radius = radius * scale, center = centre)
        }
        drawCircle(
            color = if (active) color else Sendro.textBase.copy(alpha = 0.3f),
            radius = radius,
            center = centre,
        )
    }
}

/**
 * The hatched square that stands in for a preview: the same 135-degree
 * repeating hatch the prototype uses, drawn straight into a Canvas.
 */
fun DrawScope.drawHatch(cornerRadiusPx: Float, alpha: Float = 1f) {
    val stripe = 5.dp.toPx()
    val clip = Path().apply {
        addRoundRect(
            androidx.compose.ui.geometry.RoundRect(
                left = 0f,
                top = 0f,
                right = size.width,
                bottom = size.height,
                radiusX = cornerRadiusPx,
                radiusY = cornerRadiusPx,
            ),
        )
    }
    clipPath(clip) {
        drawRect(Color.White.copy(alpha = 0.04f * alpha), size = Size(size.width, size.height))
        var x = -size.height
        while (x < size.width + size.height) {
            drawLine(
                color = Color.White.copy(alpha = 0.07f * alpha),
                start = Offset(x, 0f),
                end = Offset(x + size.height, size.height),
                strokeWidth = stripe,
            )
            x += stripe * 2
        }
    }
}
